source/base/hdf5.cc

```diff
@@ -0,0 +1,1513 @@
+    typename std::enable_if<
+      std::is_same<Container<number>, std::vector<number>>::value,
+      void *>::type
+    get_container_pointer(Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong
+      return data.data();
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, Vector<number>>::value,
+      void *>::type
+    get_container_pointer(Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &data[0];
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, FullMatrix<number>>::value,
+      void *>::type
+    get_container_pointer(Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &data[0][0];
+    }
+
+    // This function returns the pointer to the raw data of a container
+    // The returned pointer is const, which means that it can be used only to
+    // read the data
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, std::vector<number>>::value,
+      const void *>::type
+    get_container_const_pointer(const Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong
+      return data.data();
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, Vector<number>>::value,
+      const void *>::type
+    get_container_const_pointer(const Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &*data.begin();
+    }
+
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, FullMatrix<number>>::value,
+      const void *>::type
+    get_container_const_pointer(const Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &data[0][0];
+    }
+
+    // This function initializes a container of number type
```

Comment:
Can you say what "initialization" actually means in this context? What does initializing a `vector`/`Vector`/`FullMatrix` actually entail -- setting it to a concrete size, putting certain values into it?

Reply:
I added the comment:

This function initializes and returns a container of type std::vector, Vector or FullMatrix. The function does not set the values of the elements of the container. The container can store data of a HDF5 dataset or a HDF5 selection. The dimensions parameter holds the dimensions of the HDF5 dataset or selection.

In the case of a std::vector, the size of the vector will be the total size given by dimensions. For example in the case of a dataset of rank 3, the dimensions are std::vector<hsize_t>{dim_0,dim_1,dim_2}. The size of the returned std::vector will be `dim_0*dim_1*dim_2`

In the case of a dealii::Vector, the size of the returned dealii::Vector will be as well `dim_0*dim_1*dim_2`
 
 A FullMatrix can store only data of HDF5 datasets with rank 2. The size of the FullMatrix will be FullMatrix(dim_0,dim_2)